README.md
## react-native-drawer-menu [![Build Status](https://travis-ci.org/Tinysymphony/react-native-drawer-menu.svg?branch=master)](https://travis-ci.org/Tinysymphony/react-native-drawer-menu)

A drawer component for React Native Application (ios / android)

Similar to drawer menu component of QQ mobile.

### Examples

#### iOS Platform

<a href="#ios-left" id="ios-left"><img src="./GIF/ios-left-default.gif"  align="left" width="200"></a>

<a href="#ios-right" id="ios-right"><img src="./GIF/ios-right-overlay.gif" width="200"/></a>

#### Android Platform

<a href="#android-left" id="android-left"><img src="./GIF/android-left-default.gif"  align="left" width="200"></a>

<a href="#android-right" id="android-right"><img src="./GIF/android-right-overlay.gif"  width="200"></a>

### Usage

**install from npm**

``` shell
npm install --save react-native-drawer-menu
```

**import in project**

``` js
import Drawer from 'react-native-drawer-menu';

```

```js
// in render function
render() {
  // prepare your drawer content
  var drawerContent = (<View style={styles.drawerContent}>
    <View style={styles.leftTop}/>
    <View style={styles.leftBottom}>
      <View><Text>Drawer Content</Text></View>
    </View>
  </View>);
  // customize drawer's style (Optional)
  var customStyles = {
    drawer: {
      shadowColor: '#000',
      shadowOpacity: 0.4,
      shadowRadius: 10
    },
    mask: {}, // style of mask if it is enabled
    main: {} // style of main board
  };
  return (
    <Drawer
      style={styles.container}
      drawerWidth={300}
      drawerContent={drawerContent}
      type={Drawer.types.Overlay}
      customStyles={{drawer: styles.drawer}}
      drawerPosition={Drawer.positions.Right}
      onDrawerOpen={() => {console.log('Drawer is opened');}}
      onDrawerClose={() => {console.log('Drawer is closed')}}
    >
      <View style={styles.content}>
        <Text>{Object.values(Drawer.positions).join(' ')}</Text>
        <Text>{Object.values(Drawer.types).join(' ')}</Text>
      </View>
    </Drawer>
  );
}
```

> Notice: The reference of the drawer is passed to drawer content element, you could use `this.props.drawer` to invoke Drawer's instance methods like `this.props.drawer.closeDrawer()`

## Properties

| Property | Type | Default | Description |
| --- | --- | --- | --- |
| type | String | ‘default' | Type of the drawer. `default` / `overlay` You can also use static value `Drawer.type.Default` / `Drawer.type.Overlay`. |
| drawerPosition | String | ‘left' | Determine where does the drawer come out. `left` or `right` You can also use static value `Drawer.position.Left` / `Drawer.position.Right`. |
| drawerWidth | Number | 200 | The width of drawer, it’s disabled when use `replace` type. |
| duration | Number | 160 | The duration of animation to open or close drawer. |
| maskAlpha | Number | 0.4 | Maximum value is 0.5, the opactiy value of the mask over the main board when drawer is open. Mask can be disabled with `showMask` property. |
| showMask | Bool | true | Whether show the mask when drawer is open. |
| customStyles | Object | {} | Customize drawer styles. You can customize main / mask / drawer. |
| onDrawerOpen | function | null | Triggers when drawer is totally opened. |
| onDrawerClose | function | null | Triggers when drawer is totally closed. |
| startCapture | Bool | false | Whether to capture touch events while clicking on screen. |
| moveCapture | Bool | false | Whether to capture touch events while swiping over the screen. |
| responderNegotiate | function | null | Customize conditions to set pan responder, `evt` & `gestureState` will be passed as arguments. Default condition is left 20% area on screen in `left` Drawer, or right 20% area on screen in `right` Drawer. |


### Instance methods

Use ref to invoke instance methods.

| Method | Description |
| --- | --- |
| openDrawer | OpenDrawer manually |
| closeDrawer | CloseDrawer manually. The drawerContent has a ref of drawer instance, you can also trigger with it. |




CHANGELOG.md
<a name="0.0.3"></a>
## 0.0.3 (2017-02-20)


### Features

* **drawer:** Add responder negotiate function & update readme & add gif examples ([7d10be9](https://github.com/Tinysymphony/react-native-drawer-xg/commit/7d10be9))
* **Drawer:** finish overlay mode && add drawer in right position ([c46d956](https://github.com/Tinysymphony/react-native-drawer-xg/commit/c46d956))
* **Drawer/:** finish 'default' drawer in 'left' position ([750034c](https://github.com/Tinysymphony/react-native-drawer-xg/commit/750034c))



<a name="0.0.1"></a>
## 0.0.1 (2017-02-16)


### Features

* **Drawer/:** finish 'default' drawer in 'left' position ([750034c](https://github.com/Tinysymphony/react-native-drawer-xg/commit/750034c))





__tests__/test.js
/* eslint-disable */

import {
  View,
  Text,
  Dimensions
} from 'react-native';
import React from 'react';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';

import Drawer from '../Drawer';

var {width, height, scale} = Dimensions.get('window');

var drawerContent = (
  <View style={{flex: 1}}>
    <Text>test</Text>
  </View>
);

var isOpen;
let treeA = shallow(
  <Drawer
    drawerContent={drawerContent}
    duration={200}
    drawerPosition={Drawer.positions.Left}
    drawerWidth={240}
    type={Drawer.types.Default}
    onDrawerOpen={() => {isOpen = true;}}
    onDrawerClose={() => {isOpen = false;}}
  >
    <View style={{flex: 1, justifyContent: 'center', alignItems: 'center'}}>
      <Text>Hello World</Text>
    </View>
  </Drawer>
);

it('Drawer Test [Type: default, Position: left]', () => {
  let instance = treeA.instance();
  let tree = treeA;
  expect(instance.isLeft).toEqual(true);
  expect(tree.state('showMask')).toEqual(false);
  instance.openDrawer();
  setTimeout(function () {
    expect(tree.state('showMask')).toEqual(true);
    expect(isOpen).toEqual(true);
    expect(instance._onMoveShouldSetPanResponder(null, {
      dx: -10,
      dy: 0,
      moveX: 650
    })).toEqual(true);
  })
  instance.closeDrawer();
  setTimeout(function () {
    expect(tree.state('showMask')).toEqual(false);
    expect(isOpen).toEqual(false);
  }, 0);
  expect(instance._onStartShouldSetPanResponder()).toEqual(false);
  expect(instance._onMoveShouldSetPanResponder(null, {
    dx: 10,
    dy: 0,
    moveX: 10
  })).toEqual(true);
  expect(instance._onMoveShouldSetPanResponder(null, {
    dx: 10,
    dy: 0,
    moveX: 400
  })).toEqual(false);
  instance._handlePanResponderMove(null, {
    dx: 200,
  });
  expect(instance.styles.main.style.left).toEqual(200);
  instance._handlePanResponderEnd();
  setTimeout(function () {
    expect(instance.isOpen).toEqual(true);
    expect(instance.styles.main.style.left).toEqual(240);
  }, 0);
  instance.closeDrawer();
  instance._handlePanResponderMove(null, {
    dx: 100,
  });
  setTimeout(function () {
    expect(instance.isOpen).toEqual(false);
    expect(instance.styles.main.style.left).toEqual(0);
  }, 0);
});


index.ios.js
/**
 * Created by wangyi27 on 2017-02-14.
 */

import React, {Component} from 'react';
import {
  AppRegistry,
  StyleSheet,
  Text,
  View,
  Dimensions
} from 'react-native';

import Drawer from './Drawer';

const {width, height} = Dimensions.get('window');

export default class drawer extends Component {
  render() {
    var drawerContent = (<View style={styles.drawerContent}>
      <View style={styles.leftTop}/>
      <View style={styles.leftBottom}>
        <View><Text>Item1</Text></View>
        <View><Text>Item2</Text></View>
        <View><Text>Item3</Text></View>
      </View>
    </View>);
    return (
      <Drawer
        style={styles.container}
        drawerWidth={width}
        drawerContent={drawerContent}
        type={Drawer.types.Overlay}
        customStyles={{drawer: styles.drawer}}
        drawerPosition={Drawer.positions.Right}
      >
        <View style={styles.content}>
          <View style={styles.head}/>
          <Text onPress={()=>{console.log(2);}}>{width} {height}</Text>
          <Text>{Object.values(Drawer.positions).join(' ')}</Text>
          <Text>{Object.values(Drawer.types).join(' ')}</Text>
        </View>
      </Drawer>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center'
  },
  main: {
    position: 'absolute',
    backgroundColor: '#2ba'
  },
  head: {
    height: 60,
    marginBottom: 200,
    justifyContent: 'center',
    alignItems: 'center',
    alignSelf: 'stretch',
    backgroundColor: '#6a0d45'
  },
  content: {
    flex: 1,
    alignItems: 'center',
    alignSelf: 'stretch',
    backgroundColor: '#e3b8cb'
  },
  drawerContent: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center'
  },
  leftTop: {
    flex: 1,
    justifyContent: 'space-around',
    alignItems: 'stretch',
    alignSelf: 'stretch',
    backgroundColor: '#8ad8dd'
  },
  leftBottom: {
    flex: 2,
    justifyContent: 'space-around',
    alignItems: 'stretch',
    alignSelf: 'stretch',
    backgroundColor: '#f0f0f0'
  },
  drawer: {
    shadowColor: '#000',
    shadowOpacity: 0.4,
    shadowRadius: 10
  }
});

AppRegistry.registerComponent('drawer', () => drawer);


Drawer/Animation.js
/**
 * Created by TinySymphony on 2017-02-14.
 */

export default function Animation(option) {
  this.animate = this.animate.bind(this);
  this.start = this.start.bind(this);
  this.option = option;
}

Animation.prototype.animate = function (now) {
  const {
    start,
    end,
    duration,
    onAnimationFrame,
    onAnimationEnd
  } = this.option;
  var currentDuration = now - this.startTime;
  if (currentDuration >= duration) {
    onAnimationFrame(end);
    onAnimationEnd();
    return;
  }
  let value;
  // TODO more animation function
  if (start > end) {
    value = start - (start - end) * currentDuration / duration;
  } else {
    value = (end - start) * currentDuration / duration + start;
  }
  onAnimationFrame(value);
  requestAnimationFrame(this.animate);
};

Animation.prototype.start = function () {
  this.startTime = new Date();
  this.animate(this.startTime);
};


ios/drawer/AppDelegate.m
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

#import "AppDelegate.h"

#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  NSURL *jsCodeLocation;

  jsCodeLocation = [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index.ios" fallbackResource:nil];

  RCTRootView *rootView = [[RCTRootView alloc] initWithBundleURL:jsCodeLocation
                                                      moduleName:@"drawer"
                                               initialProperties:nil
                                                   launchOptions:launchOptions];
  rootView.backgroundColor = [[UIColor alloc] initWithRed:1.0f green:1.0f blue:1.0f alpha:1];

  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [UIViewController new];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];
  return YES;
}

@end


ios/drawerTests/drawerTests.m
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

#import <UIKit/UIKit.h>
#import <XCTest/XCTest.h>

#import <React/RCTLog.h>
#import <React/RCTRootView.h>

#define TIMEOUT_SECONDS 600
#define TEXT_TO_LOOK_FOR @"Welcome to React Native!"

@interface drawerTests : XCTestCase

@end

@implementation drawerTests

- (BOOL)findSubviewInView:(UIView *)view matching:(BOOL(^)(UIView *view))test
{
  if (test(view)) {
    return YES;
  }
  for (UIView *subview in [view subviews]) {
    if ([self findSubviewInView:subview matching:test]) {
      return YES;
    }
  }
  return NO;
}

- (void)testRendersWelcomeScreen
{
  UIViewController *vc = [[[[UIApplication sharedApplication] delegate] window] rootViewController];
  NSDate *date = [NSDate dateWithTimeIntervalSinceNow:TIMEOUT_SECONDS];
  BOOL foundElement = NO;

  __block NSString *redboxError = nil;
  RCTSetLogFunction(^(RCTLogLevel level, RCTLogSource source, NSString *fileName, NSNumber *lineNumber, NSString *message) {
    if (level >= RCTLogLevelError) {
      redboxError = message;
    }
  });

  while ([date timeIntervalSinceNow] > 0 && !foundElement && !redboxError) {
    [[NSRunLoop mainRunLoop] runMode:NSDefaultRunLoopMode beforeDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
    [[NSRunLoop mainRunLoop] runMode:NSRunLoopCommonModes beforeDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];

    foundElement = [self findSubviewInView:vc.view matching:^BOOL(UIView *view) {
      if ([view.accessibilityLabel isEqualToString:TEXT_TO_LOOK_FOR]) {
        return YES;
      }
      return NO;
    }];
  }

  RCTSetLogFunction(RCTDefaultLogFunction);

  XCTAssertNil(redboxError, @"RedBox error: %@", redboxError);
  XCTAssertTrue(foundElement, @"Couldn't find element with text '%@' in %d seconds", TEXT_TO_LOOK_FOR, TIMEOUT_SECONDS);
}


@end


android/app/src/main/java/com/drawer/MainApplication.java
package com.drawer;

import android.app.Application;
import android.util.Log;

import com.facebook.react.ReactApplication;
import com.facebook.react.ReactInstanceManager;
import com.facebook.react.ReactNativeHost;
import com.facebook.react.ReactPackage;
import com.facebook.react.shell.MainReactPackage;
import com.facebook.soloader.SoLoader;

import java.util.Arrays;
import java.util.List;

public class MainApplication extends Application implements ReactApplication {

  private final ReactNativeHost mReactNativeHost = new ReactNativeHost(this) {
    @Override
    public boolean getUseDeveloperSupport() {
      return BuildConfig.DEBUG;
    }

    @Override
    protected List<ReactPackage> getPackages() {
      return Arrays.<ReactPackage>asList(
          new MainReactPackage()
      );
    }
  };

  @Override
  public ReactNativeHost getReactNativeHost() {
    return mReactNativeHost;
  }

  @Override
  public void onCreate() {
    super.onCreate();
    SoLoader.init(this, /* native exopackage */ false);
  }
}


Drawer/index.js
/**
 * Created by TinySymphony on 2017-02-14.
 */

import React, {PropTypes, Component} from 'react';
import {
  StyleSheet,
  View,
  PanResponder,
  Dimensions
} from 'react-native';

const {width} = Dimensions.get('window');

import Animation from './Animation';

/* eslint-disable curly, no-new, no-warning-comments */
const positions = {
  Left: 'left',
  Right: 'right'
};

const types = {
  Overlay: 'overlay',
  Default: 'default',
  Replace: 'replace'
};

const MAX = {
  maskAlpha: 0.5
};

export default class Drawer extends Component {
  static positions = positions
  static types = types
  static defaultProps = {
    drawerWidth: 200,
    duration: 160,
    drawerPosition: positions.Left,
    type: types.Default,
    showMask: true,
    maskAlpha: 0.4,
    customStyles: {},
    startCapture: false,
    moveCapture: false
  }
  static propTypes = {
    drawerContent: PropTypes.object,
    width: PropTypes.number,
    duration: PropTypes.number,
    drawerPosition: PropTypes.oneOf(Object.values(positions)),
    type: PropTypes.oneOf(Object.values(types)),
    showMask: PropTypes.bool,
    maskAlpha: PropTypes.number,
    customStyles: PropTypes.object,
    onDrawerClose: PropTypes.func,
    onDrawerOpen: PropTypes.func,
    startCapture: PropTypes.bool,
    moveCapture: PropTypes.bool,
    responderNegotiate: PropTypes.func
  }
  componentWillMount() {
    const {
      drawerWidth,
      drawerPosition,
      maskAlpha
    } = this.props;
    this.isLeft = drawerPosition === positions.Left;
    this.MAX_DX = drawerWidth > 0.8 * width ? 0.8 * width : drawerWidth;
    this.MAX_ALPHA = maskAlpha > MAX.maskAlpha ? MAX.maskAlpha : maskAlpha;
    this.styles = {
      drawer: {
        style: {
          top: 0,
          bottom: 0,
          left: this.isLeft ? -this.MAX_DX : width,
          right: this.isLeft ? width : -this.MAX_DX
        }
      },
      main: {
        style: {
          top: 0,
          bottom: 0,
          left: 0,
          right: 0
        }
      },
      mask: {
        style: {
          backgroundColor: 'rgba(0, 0, 0, 0)'
        }
      }
    };
    this.inAnimation = false;
    this.isOpen = false;
    this._pan = PanResponder.create({
      onStartShouldSetPanResponder: this._onStartShouldSetPanResponder.bind(this),
      onStartShouldSetPanResponderCapture: (evt, gestureState) => this.props.startCapture,
      onMoveShouldSetPanResponder: this._onMoveShouldSetPanResponder.bind(this),
      onMoveShouldSetPanResponderCapture: (evt, gestureState) => this.props.moveCapture,
      onPanResponderTerminationRequest: (evt, gestureState) => true,
      onPanResponderGrant: this._handlePanResponderGrant.bind(this),
      onPanResponderMove: this._handlePanResponderMove.bind(this),
      onPanResponderRelease: this._handlePanResponderEnd.bind(this),
      onPanResponderTerminate: this._handlePanResponderEnd.bind(this),
      onShouldBlockNativeResponder: (evt, gestureState) => true
    });
    this.state = {
      showMask: false
    };
    this._handleMainBoardPress = this._handleMainBoardPress.bind(this);
    this._drawerDidClose = this._drawerDidClose.bind(this);
    this._drawerDidOpen = this._drawerDidOpen.bind(this);
  }
  componentDidMount() {
    this._updateNativeStyles(0);
  }
  _onStartShouldSetPanResponder (evt, gestureState) {
    // set responder for tapping when the drawer is open
    if (this.isOpen && !this.inAnimation) return true;
    return false;
  }
  _onMoveShouldSetPanResponder (evt, gestureState) {
    // custom pan responder condition function
    if (this.props.responderNegotiate && this.props.responderNegotiate(evt, gestureState) === false) return false;
    if (this._touchPositionCheck(gestureState)) {
      this.props.showMask && !this.state.showMask && this.setState({showMask: true});
      // this.props.onDrawerStartOpen && this.props.onDrawerStartOpen();
      return true;
    }
    return false;
  }
  _handlePanResponderGrant(evt, gestureState) {
  }
  _handlePanResponderMove (evt, gestureState) {
    let dx = gestureState.dx;
    if (dx > 0 && dx <= this.MAX_DX) {
      // swipe right
      if (this.isOpen && !this.isLeft) this._updateNativeStyles(-this.MAX_DX + dx);
      if (!this.isOpen && this.isLeft) this._updateNativeStyles(dx);
    } else if (dx < 0 && dx >= -this.MAX_DX) {
      // swipe left
      if (this.isOpen && this.isLeft) this._updateNativeStyles(this.MAX_DX + dx);
      if (!this.isOpen && !this.isLeft) this._updateNativeStyles(dx);
    }
    // dx === 0 triggers tap event when drawer is opened.
  }
  _handlePanResponderEnd (evt, gestureState) {
    let currentWidth = this._getCurrentDrawerWidth();
    if (!this.isLeft) currentWidth *= -1;
    if (this.isOpen && gestureState.dx === 0) return this._handleMainBoardPress();
    if (currentWidth === this.MAX_DX) return this._drawerDidOpen();
    if (currentWidth === 0) return this._drawerDidClose();
    if (currentWidth > this.MAX_DX / 2) {
      this.openDrawer();
    } else {
      this.closeDrawer();
    }
  }
  _getCurrentDrawerWidth () {
    return this.isLeft ? this.styles.drawer.style.left + this.MAX_DX :
      this.styles.drawer.style.left - width;
  }
  _touchPositionCheck(gestureState) {
    const {moveX, dx, dy} = gestureState;
    let x0 = moveX; // in move set panresponder state, moveX is the original point's coordinates
    let result = false;
    if (Math.abs(dx) < Math.abs(dy)) return result;
    if (this.isOpen && dx !== 0 ||
      this.isLeft && x0 <= width * 0.2 && !this.isOpen && dx > 0 ||
      !this.isLeft && x0 >= width * 0.8 && !this.isOpen && dx < 0) {
      return true;
    }
    return false;
  }
  closeDrawer() {
    if (this.inAnimation) return;
    this.inAnimation = true;
    const {duration} = this.props;
    let left = this._getCurrentDrawerWidth();
    new Animation({
      start: left,
      end: 0,
      duration,
      onAnimationFrame: (val) => {
        this._updateNativeStyles(val);
      },
      onAnimationEnd: this._drawerDidClose
    }).start();
  }
  openDrawer() {
    if (this.inAnimation) return;
    this.inAnimation = true;
    const {duration} = this.props;
    let left = this._getCurrentDrawerWidth();
    this.props.showMask && !this.state.showMask && this.setState({showMask: true});
    new Animation({
      start: left,
      end: this.isLeft ? this.MAX_DX : -this.MAX_DX,
      duration,
      onAnimationFrame: (val) => {
        this._updateNativeStyles(val);
      },
      onAnimationEnd: this._drawerDidOpen
    }).start();
  }
  _drawerDidOpen () {
    this.inAnimation = false;
    this.isOpen = true;
    this.props.onDrawerOpen && this.props.onDrawerOpen();
  }
  _drawerDidClose () {
    this.inAnimation = false;
    this.isOpen = false;
    this.state.showMask && this.setState({showMask: false}, () => {
      this.props.onDrawerClose && this.props.onDrawerClose();
    });
  }
  _updateNativeStyles (dx) {
    this.styles.drawer.style.left = this.isLeft ? -this.MAX_DX + dx : width + dx;
    this.styles.drawer.style.right = this.isLeft ? width - dx : -this.MAX_DX - dx;
    this.styles.mask.style.backgroundColor = `rgba(0, 0, 0,
      ${(Math.abs(dx) / this.MAX_DX * this.MAX_ALPHA).toFixed(2)})`;
    this._drawer && this._drawer.setNativeProps(this.styles.drawer);
    this._mask && this._mask.setNativeProps(this.styles.mask);
    if (this.props.type === types.Default || dx === 0) {
      this.styles.main.style.left = dx;
      this.styles.main.style.right = -dx;
      this._main && this._main.setNativeProps(this.styles.main);
    }
  }
  _handleMainBoardPress () {
    if (this.inAnimation) return;
    this.closeDrawer();
  }
  render() {
    const {customStyles} = this.props;
    let drawerContent;
    if (this.props.drawerContent) {
      drawerContent = React.cloneElement(this.props.drawerContent, {
        drawer: this
      });
    }
    return (
      <View style={styles.container}>
        <View
          ref={(main) => {this._main = main;}}
          style={[customStyles.main, styles.absolute]}
          {...this._pan.panHandlers}
        >
          {this.props.children}
          {this.state.showMask && <View
            ref={(mask) => {this._mask = mask;}}
            style={[customStyles.mask, styles.mask, styles.absolute]}/>}
        </View>
        <View ref={(drawer) => {this._drawer = drawer;}} style={[customStyles.drawer, styles.absolute]}>
          {drawerContent}
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1
  },
  absolute: {
    position: 'absolute'
  },
  mask: {
    left: 0,
    top: 0,
    bottom: 0,
    right: 0
  }
});
